sdk/storage/azure-storage-datamovement/src/job_engine_serialization.cpp

```diff
@@ -0,0 +1,595 @@
+        {
+          if (!doneBits[currDoneBit])
+          {
+            auto task = TaskModel::FromString(ReadString(fin));
+            tasks.push_back(std::move(task));
+            ++numUndoneBits;
+          }
+        }
+        else
+        {
+          bool hasUndoneSubtask = false;
+          std::string subtasksDoneBitMap(numSubtasks, '\x00');
+          for (size_t i = 0; i < static_cast<size_t>(numSubtasks); ++i)
+          {
+            subtasksDoneBitMap[i] = char(doneBits[currDoneBit + i] + '0');
+            if (subtasksDoneBitMap[i] == '0')
+            {
+              hasUndoneSubtask = true;
+              ++numUndoneBits;
+            }
+          }
+          if (hasUndoneSubtask)
+          {
+            auto task = TaskModel::FromString(ReadString(fin));
+            task.ExtendedAttributes["_subtasks"] = subtasksDoneBitMap;
+            tasks.push_back(std::move(task));
+          }
+        }
+        currDoneBit += numSubtasks;
+      }
+      fin.close();
+      jobPart.m_mappedFile = std::make_unique<_internal::MemoryMap>(partFilename);
+      jobPart.m_doneBitmap = static_cast<bool*>(
+          jobPart.m_mappedFile->Map(g_JobPartFileHeaderSize, jobPart.m_numDoneBits));
+      jobPart.m_numUndoneBits->store(numUndoneBits, std::memory_order_relaxed);
+      return std::make_pair(std::move(jobPart), std::move(tasks));
+    }
+
+    void JobPart::CreateJobPart(
+        uint32_t id,
+        const std::string& jobPlanDir,
+        const std::vector<TaskModel>& tasks)
+    {
+      const std::string partFilename = _internal::JoinPath(jobPlanDir, PartIdToString(id));
+      std::fstream fOut(
+          partFilename + ".tmp", std::fstream::out | std::fstream::trunc | std::fstream::binary);
+      fOut.exceptions(std::fstream::failbit | std::fstream::badbit);
+      WriteFixedInt(fOut, g_PlanFileVersion);
+      int32_t numDoneBits
+          = std::accumulate(tasks.begin(), tasks.end(), 0, [](int32_t s, const TaskModel& t) {
+              return s + t.NumSubtasks;
+            });
+      WriteFixedInt(fOut, numDoneBits);
+      WriteZeros(fOut, numDoneBits);
+      for (const auto& t : tasks)
+      {
+        WriteVarInt(fOut, t.NumSubtasks);
+        WriteString(fOut, t.ToString());
+      }
+      fOut.close();
+      _internal::Rename(partFilename + ".tmp", partFilename);
+    }
+
+    void JobPlan::CreateJobPlan(const _internal::JobModel& model, const std::string& jobPlanDir)
+    {
+      AZURE_ASSERT(!_internal::PathExists(jobPlanDir));
+      _internal::CreateDirectory(jobPlanDir);
+      {
+        const std::string partGensFilename = _internal::JoinPath(jobPlanDir, "part_gens");
+        std::fstream fOut(
+            partGensFilename + ".tmp",
+            std::fstream::out | std::fstream::trunc | std::fstream::binary);
+        fOut.exceptions(std::fstream::failbit | std::fstream::badbit);
+        WriteFixedInt(fOut, g_PlanFileVersion);
+        PartGenerator rootGenerator;
```

Comment:
rootGenerator is not assigned any value, just serialized its default value to disk?

Reply:
Yes, empty means this is root generator.